Add support for service level auto_pause_notifications_parameters

pdecat opened the pull request:
This PR adds support for service level `auto_pause_notifications_parameters`.

Dependencies were updated with `go get -u ./... && go mod vendor && go mod tidy` to fetch updates from:
* https://github.com/heimweh/go-pagerduty/pull/92
* https://github.com/heimweh/go-pagerduty/pull/93

Acceptance test results:

```
make testacc TEST=./pagerduty/ TESTARGS='-run=TestAccPagerDutyService_AutoPauseNotificationsParameters'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./pagerduty/ -v -run=TestAccPagerDutyService_AutoPauseNotificationsParameters -timeout 120m
=== RUN   TestAccPagerDutyService_AutoPauseNotificationsParameters
--- PASS: TestAccPagerDutyService_AutoPauseNotificationsParameters (20.23s)
PASS
ok      github.com/terraform-providers/terraform-provider-pagerduty/pagerduty   20.240s
```

pdecat commented on pagerduty/resource_pagerduty_service_test.go:
```diff
@@ -242,6 +242,83 @@ func TestAccPagerDutyService_AlertContentGrouping(t *testing.T) {
+	service := fmt.Sprintf("tf-%s", acctest.RandString(5))
+
+	resource.Test(t, resource.TestCase{
+		// PreCheck:     func() { testAccPreCheck(t); testAccPreCheckPagerDutyAbility(t, "XXX") }, // FIXME: not sure which ability this feature is depending on https://developer.pagerduty.com/api-reference/9def78ed82b74-list-abilities
```
I am not sure which ability this feature is depending on https://developer.pagerduty.com/api-reference/9def78ed82b74-list-abilities

pdecat commented:
Maybe vendored dependencies should be dropped altogether to only rely on `go.sum`, and avoid having so many files impacted on dependencies updates.

pdecat commented:
Marked as draft as I missed some important parts in my former commit, and I'm now facing issues.

pdecat commented:
This is now ready for review.

Acceptance test results:

```
make testacc TEST=./pagerduty/ TESTARGS='-run=TestAccPagerDutyService_AutoPauseNotificationsParameters'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./pagerduty/ -v -run=TestAccPagerDutyService_AutoPauseNotificationsParameters -timeout 120m
=== RUN   TestAccPagerDutyService_AutoPauseNotificationsParameters
--- PASS: TestAccPagerDutyService_AutoPauseNotificationsParameters (19.31s)
PASS
ok      github.com/terraform-providers/terraform-provider-pagerduty/pagerduty   19.323s
```

pdecat commented on pagerduty/resource_pagerduty_service_test.go:
```diff
@@ -242,6 +242,87 @@ func TestAccPagerDutyService_AlertContentGrouping(t *testing.T) {
+					resource.TestCheckResourceAttr(
+						"pagerduty_service.foo", "auto_pause_notifications_parameters.0.enabled", "false"),
+					resource.TestCheckResourceAttr(
+						"pagerduty_service.foo", "auto_pause_notifications_parameters.0.timeout", "0"),
```
We get 0 instead of `null` because that's the default for TypeInt, see https://github.com/hashicorp/terraform-plugin-sdk/issues/90.

Switching to https://github.com/hashicorp/terraform-plugin-framework in the future may help here.

pdecat commented:
Rebased on master, acceptance tests results:

```
make testacc TEST=./pagerduty/ TESTARGS='-run=TestAccPagerDutyService_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./pagerduty/ -v -run=TestAccPagerDutyService_ -timeout 120m
=== RUN   TestAccPagerDutyService_import
--- PASS: TestAccPagerDutyService_import (12.39s)
=== RUN   TestAccPagerDutyService_Basic
--- PASS: TestAccPagerDutyService_Basic (19.22s)
=== RUN   TestAccPagerDutyService_AlertGrouping
--- PASS: TestAccPagerDutyService_AlertGrouping (16.46s)
=== RUN   TestAccPagerDutyService_AlertContentGrouping
    resource_pagerduty_service_test.go:206: Step 2/2 error: Check failed: Check 8/12 error: pagerduty_service.foo: list or set attribute 'alert_grouping_parameters.0.config' must be checked by element count key (alert_grouping_parameters.0.config.#) or element value keys (e.g. alert_grouping_parameters.0.config.0). Set element value checks should use TestCheckTypeSet functions instead.
--- FAIL: TestAccPagerDutyService_AlertContentGrouping (13.98s)
=== RUN   TestAccPagerDutyService_AutoPauseNotificationsParameters
--- PASS: TestAccPagerDutyService_AutoPauseNotificationsParameters (18.85s)
=== RUN   TestAccPagerDutyService_BasicWithIncidentUrgencyRules
--- PASS: TestAccPagerDutyService_BasicWithIncidentUrgencyRules (21.10s)
=== RUN   TestAccPagerDutyService_FromBasicToCustomIncidentUrgencyRules
--- PASS: TestAccPagerDutyService_FromBasicToCustomIncidentUrgencyRules (15.91s)
=== RUN   TestAccPagerDutyService_SupportHoursChange
--- PASS: TestAccPagerDutyService_SupportHoursChange (15.53s)
FAIL
FAIL    github.com/terraform-providers/terraform-provider-pagerduty/pagerduty   133.449s
FAIL
make: *** [GNUmakefile:17: testacc] Error 1
```

The `TestAccPagerDutyService_AlertContentGrouping` failure is caused by the updated terraform SDK that performs additional checks. I'll submit a fix.

pdecat commented:
Rebased on master and pulled in https://github.com/heimweh/go-pagerduty/commit/5d470ed7c5fc9141265a7ba6590c027e30f1d46c, acceptance tests results:

```
# make testacc TEST=./pagerduty/ TESTARGS='-run=TestAccPagerDutyService_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./pagerduty/ -v -run=TestAccPagerDutyService_ -timeout 120m
=== RUN   TestAccPagerDutyService_import
--- PASS: TestAccPagerDutyService_import (12.67s)
=== RUN   TestAccPagerDutyService_Basic
--- PASS: TestAccPagerDutyService_Basic (19.62s)
=== RUN   TestAccPagerDutyService_AlertGrouping
--- PASS: TestAccPagerDutyService_AlertGrouping (15.56s)
=== RUN   TestAccPagerDutyService_AlertContentGrouping
    resource_pagerduty_service_test.go:206: Step 2/2 error: Check failed: Check 8/12 error: pagerduty_service.foo: list or set attribute 'alert_grouping_parameters.0.config' must be checked by element count key (alert_grouping_parameters.0.config.#) or element value keys (e.g. alert_grouping_parameters.0.config.0). Set element value checks should use TestCheckTypeSet functions instead.
--- FAIL: TestAccPagerDutyService_AlertContentGrouping (14.85s)
=== RUN   TestAccPagerDutyService_AutoPauseNotificationsParameters
--- PASS: TestAccPagerDutyService_AutoPauseNotificationsParameters (19.02s)
=== RUN   TestAccPagerDutyService_BasicWithIncidentUrgencyRules
--- PASS: TestAccPagerDutyService_BasicWithIncidentUrgencyRules (20.84s)
=== RUN   TestAccPagerDutyService_FromBasicToCustomIncidentUrgencyRules
--- PASS: TestAccPagerDutyService_FromBasicToCustomIncidentUrgencyRules (16.19s)
=== RUN   TestAccPagerDutyService_SupportHoursChange
--- PASS: TestAccPagerDutyService_SupportHoursChange (15.84s)
FAIL
FAIL    github.com/terraform-providers/terraform-provider-pagerduty/pagerduty   134.603s
FAIL
make: *** [GNUmakefile:17: testacc] Error 1
```

pdecat commented:
Hi @imjaroiswebdev, the alert grouping acceptance test is still triggering errors with the new terraform SDK:

```
# make testacc TEST=./pagerduty/ TESTARGS='-run=TestAccPagerDutyService_AlertContentGrouping'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./pagerduty/ -v -run=TestAccPagerDutyService_AlertContentGrouping -timeout 120m
=== RUN   TestAccPagerDutyService_AlertContentGrouping
    resource_pagerduty_service_test.go:206: Step 2/2 error: Check failed: Check 8/12 error: pagerduty_service.foo: list or set attribute 'alert_grouping_parameters.0.config' must be checked by element count key (alert_grouping_parameters.0.config.#) or element value keys (e.g. alert_grouping_parameters.0.config.0). Set element value checks should use TestCheckTypeSet functions instead.
--- FAIL: TestAccPagerDutyService_AlertContentGrouping (14.83s)
FAIL
FAIL    github.com/terraform-providers/terraform-provider-pagerduty/pagerduty   14.839s
FAIL
make: *** [GNUmakefile:17: testacc] Error 1
```

Changing this to:
```
git diff
diff --git a/pagerduty/resource_pagerduty_service_test.go b/pagerduty/resource_pagerduty_service_test.go
index a01fa7cd..f17d8101 100644
--- a/pagerduty/resource_pagerduty_service_test.go
+++ b/pagerduty/resource_pagerduty_service_test.go
@@ -254,8 +254,8 @@ func TestAccPagerDutyService_AlertContentGrouping(t *testing.T) {
                                                "pagerduty_service.foo", "alert_creation", "create_alerts_and_incidents"),
                                        resource.TestCheckResourceAttr(
                                                "pagerduty_service.foo", "alert_grouping", "rules"),
-                                       resource.TestCheckNoResourceAttr(
-                                               "pagerduty_service.foo", "alert_grouping_parameters.0.config"),
+                                       resource.TestCheckResourceAttr(
+                                               "pagerduty_service.foo", "alert_grouping_parameters.0.config.#", "0"),
                                        resource.TestCheckNoResourceAttr(
                                                "pagerduty_service.foo", "alert_grouping_parameters.0.type"),
                                        resource.TestCheckResourceAttr(
```

resolves the issue:
```
# make testacc TEST=./pagerduty/ TESTARGS='-run=TestAccPagerDutyService_AlertContentGrouping'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./pagerduty/ -v -run=TestAccPagerDutyService_AlertContentGrouping -timeout 120m
=== RUN   TestAccPagerDutyService_AlertContentGrouping
--- PASS: TestAccPagerDutyService_AlertContentGrouping (15.62s)
PASS
ok      github.com/terraform-providers/terraform-provider-pagerduty/pagerduty   15.635s
```


See https://github.com/PagerDuty/terraform-provider-pagerduty/pull/525#issuecomment-1193081644

imjaroiswebdev commented:
Hi @pdecat thank you for this awesome addition to support Auto Pause Notifications Parameters, I've been reviewing it and looks great, but if you don't mind We would prefer to let the module vendoring and tidy for another PR like #532 which you already opened and focus this one in adding this improvement to the Provider. So thinking on that I have rebased and reapplied your commit in a new branch on my fork for you to see how We can do add the change without updating vendors and tidy and continue from there. Please take a look at it in the following branches comparison... https://github.com/pdecat/terraform-provider-pagerduty/compare/auto_pause_notifications_parameters...imjaroiswebdev:terraform-provider-pagerduty:auto_pause_notifications_parameters

The tests you added are also looking very good...
![Screen Shot 2022-07-28 at 14 37 04](https://user-images.githubusercontent.com/24704624/182254145-d10c6618-9dad-4a2d-96b5-032f185700ce.png)

pdecat commented:
Hi @imjaroiswebdev, updating at least part of the dependencies is required because of https://github.com/heimweh/go-pagerduty/commit/2e1f76487c18b87ddf06153f86faab7193b3a206

If you don't, `go mod vendor` complains:

```
# go get github.com/heimweh/go-pagerduty@5d470ed
go: upgraded github.com/heimweh/go-pagerduty v0.0.0-20220527195341-4e587aa9b58e => v0.0.0-20220722230120-5d470ed7c5fc

# go mod vendor
go: github.com/heimweh/go-pagerduty@v0.0.0-20220722230120-5d470ed7c5fc requires
        go.mongodb.org/mongo-driver@v1.9.1: missing go.sum entry; to add it:
        go mod download go.mongodb.org/mongo-driver
go: github.com/heimweh/go-pagerduty@v0.0.0-20220722230120-5d470ed7c5fc requires
        go.mongodb.org/mongo-driver@v1.9.1: missing go.sum entry; to add it:
        go mod download go.mongodb.org/mongo-driver
```

pdecat commented:
Hi @imjaroiswebdev, I've now rebased on master following your go modules updates, and dropped my corresponding commits from this branch.

Acceptance tests results:
```
# make testacc TEST=./pagerduty/ TESTARGS='-run=TestAccPagerDutyService_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./pagerduty/ -v -run=TestAccPagerDutyService_ -timeout 120m
=== RUN   TestAccPagerDutyService_import
--- PASS: TestAccPagerDutyService_import (13.13s)
=== RUN   TestAccPagerDutyService_Basic
--- PASS: TestAccPagerDutyService_Basic (19.95s)
=== RUN   TestAccPagerDutyService_AlertGrouping
--- PASS: TestAccPagerDutyService_AlertGrouping (16.48s)
=== RUN   TestAccPagerDutyService_AlertContentGrouping
--- PASS: TestAccPagerDutyService_AlertContentGrouping (16.27s)
=== RUN   TestAccPagerDutyService_AutoPauseNotificationsParameters
--- PASS: TestAccPagerDutyService_AutoPauseNotificationsParameters (19.37s)
=== RUN   TestAccPagerDutyService_BasicWithIncidentUrgencyRules
--- PASS: TestAccPagerDutyService_BasicWithIncidentUrgencyRules (21.66s)
=== RUN   TestAccPagerDutyService_FromBasicToCustomIncidentUrgencyRules
--- PASS: TestAccPagerDutyService_FromBasicToCustomIncidentUrgencyRules (15.50s)
=== RUN   TestAccPagerDutyService_SupportHoursChange
--- PASS: TestAccPagerDutyService_SupportHoursChange (16.09s)
=== RUN   TestAccPagerDutyService_ResponsePlay
--- PASS: TestAccPagerDutyService_ResponsePlay (12.53s)
PASS
ok      github.com/terraform-providers/terraform-provider-pagerduty/pagerduty   150.997s
```

Note: I had to run `make fmt` to fix an issue with ``:

```
# make test
==> Checking that code complies with gofmt requirements...
gofmt needs running on the following files:
./pagerduty/resource_pagerduty_slack_connection.go
You can use the command: `make fmt` to reformat code.
make: *** [GNUmakefile:32: fmtcheck] Error 1

# make fmt
gofmt -w $(find . -name '*.go' |grep -v vendor)

# make test
==> Checking that code complies with gofmt requirements...
go test -i $(go list ./... |grep -v 'vendor') || exit 1
go: -i flag is deprecated
echo $(go list ./... |grep -v 'vendor') | \
        xargs -t -n4 go test  -timeout=30s -parallel=4
go test '-timeout=30s' '-parallel=4' github.com/terraform-providers/terraform-provider-pagerduty github.com/terraform-providers/terraform-provider-pagerduty/pagerduty
?       github.com/terraform-providers/terraform-provider-pagerduty     [no test files]
ok      github.com/terraform-providers/terraform-provider-pagerduty/pagerduty   0.019s
```

imjaroiswebdev reviewed (approved):
This a great addition @pdecat thank you so much for your contribution🎉💪🏽, since you have done all the heavy lifting with this excellent PR and all the current and new tests are passing, in the end I only added the final and superficial touches for be able to merged adding a link to the docs which mention the plan restrictions for the PagerDuty account in the documentation you added, on top of that, I removed your inquire about the ability to do the precheck on the test and added the generic test precheck We have, since this configuration can be done by, but the limitation mentioned is related to its use, because of that there is no an explicit ability listed.
